File: src/main/java/com/gitlab/daring/image/util/CacheUtils.java
package com.gitlab.daring.image.util;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;

import static com.gitlab.daring.image.util.CommonUtils.closeQuietly;

public class CacheUtils {

    public static <K, V extends AutoCloseable> Cache<K, V> buildClosableCache(String spec) {
        return CacheBuilder.from(spec)
            .<K, V>removalListener(n -> closeQuietly(n.getValue()))
            .build();
    }

    private CacheUtils() {
    }

}

File: src/main/kotlin/com/gitlab/daring/image/util/CacheUtils.kt
package com.gitlab.daring.image.util

import com.gitlab.daring.image.util.CommonUtils.closeQuietly
import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder

object CacheUtils {

    @JvmStatic
    fun <K, V : AutoCloseable> buildClosableCache(spec: String): Cache<K, V> {
        return CacheBuilder.from(spec)
                .removalListener<K, V> { n -> closeQuietly(n.value) }
                .build<K, V>()
    }

}